runtime/src/accounts_db.rs

```diff
@@ -378,35 +391,34 @@ impl AccountsDB {
         Self::load(&storage, ancestors, &accounts_index, pubkey)
     }
 
-    fn fork_storage(&self, fork_id: Fork) -> Arc<AccountStorageEntry> {
-        let mut candidates: Vec<Arc<AccountStorageEntry>> = {
-            let stores = self.storage.read().unwrap();
-            let fork_stores = stores.0.get(&fork_id);
-            if let Some(fork_stores) = fork_stores {
-                fork_stores
-                    .values()
-                    .filter_map(|x| {
-                        if x.status() == AccountStorageStatus::StorageAvailable {
-                            Some(x.clone())
-                        } else {
-                            None
-                        }
-                    })
-                    .collect()
-            } else {
-                vec![]
+    fn find_storage_candidate(&self, fork_id: Fork) -> Arc<AccountStorageEntry> {
+        let stores = self.storage.read().unwrap();
+
+        if let Some(fork_stores) = stores.0.get(&fork_id) {
+            if !fork_stores.is_empty() {
+                // pick an available store at random by iterating from a random point
+                let to_skip = thread_rng().gen_range(0, stores.0.len());
+
+                for (i, store) in fork_stores.values().cycle().skip(to_skip).enumerate() {
+                    if store.try_available() {
+                        return store.clone();
+                    }
+                    // looked at every store, bail...
+                    if i == fork_stores.len() {
+                        break;
+                    }
+                }
             }
-        };
-        if candidates.is_empty() {
-            let mut stores = self.storage.write().unwrap();
-            let path_index = thread_rng().gen_range(0, self.paths.len());
-            let storage = Arc::new(self.new_storage_entry(fork_id, &self.paths[path_index]));
-            let fork_storage = stores.0.entry(fork_id).or_insert_with(HashMap::new);
-            fork_storage.insert(storage.id, storage.clone());
-            candidates.push(storage);
         }
-        let rv = thread_rng().gen_range(0, candidates.len());
-        candidates[rv].clone()
+        drop(stores);
+
+        let mut stores = self.storage.write().unwrap();
```

Comment:
Should we limit the number of entries we create here and force threads to go through another iteration of `try_available` once that limit has been reached? It is possible that as thread A is iterating and calling `try_available` on the stores above, other threads are also grabbing the locks for those stores while releasing stores that thread A has already iterated past.

Reply:
I'd prefer to have them at contention+1 ;)

Reply:
Hmmm could we add a metric to figure out how many storageentries we are creating per bank? It's not exactly equivalent because given N threads and some M > N AppendVecs, your chances of finding a free/available AppendVec should increase with more iterations of the check (probability that the scheduling screws you should decrease).